hack/generator/pkg/astmodel/conversion_function_builder.go

```diff
@@ -0,0 +1,494 @@
+	// TODO: Better way to let you fuss with this? How can you pick out what I've already put in here to overwrite it?
+	conversions []ConversionHandler
+
+	IdFactory             IdentifierFactory
+	CodeGenerationContext *CodeGenerationContext
+}
+
+// NewConversionFunctionBuilder creates a new ConversionFunctionBuilder with the default conversions already added.
+func NewConversionFunctionBuilder(idFactory IdentifierFactory, codeGenerationContext *CodeGenerationContext) *ConversionFunctionBuilder {
+	return &ConversionFunctionBuilder{
+		IdFactory:             idFactory,
+		CodeGenerationContext: codeGenerationContext,
+		conversions: []ConversionHandler{
+			// Complex wrapper types checked first
+			IdentityConvertComplexOptionalProperty,
+			IdentityConvertComplexArrayProperty,
+			IdentityConvertComplexMapProperty,
+
+			// TODO: a flip function of some kind would be kinda nice (for source vs dest)
+			IdentityAssignValidatedTypeDestination,
+			IdentityAssignValidatedTypeSource,
+			IdentityAssignPrimitiveType,
+			IdentityDeepCopyJSON,
+			IdentityAssignTypeName,
+		},
+	}
+}
+
+// AddConversionHandlers adds the specified conversion handlers to the end of the conversion list.
+func (builder *ConversionFunctionBuilder) AddConversionHandlers(conversionHandlers ...ConversionHandler) {
+	builder.conversions = append(builder.conversions, conversionHandlers...)
+}
+
+// PrependConversionHandlers adds the specified conversion handlers to the beginning of the conversion list.
+func (builder *ConversionFunctionBuilder) PrependConversionHandlers(conversionHandlers ...ConversionHandler) {
+	builder.conversions = append(conversionHandlers, builder.conversions...)
+}
+
+func (builder *ConversionFunctionBuilder) BuildConversion(params ConversionParameters) []dst.Stmt {
+	for _, conversion := range builder.conversions {
+		result := conversion(builder, params)
+		if len(result) > 0 {
+			return result
+		}
+	}
+
+	panic(fmt.Sprintf("don't know how to perform conversion for %T -> %T", params.SourceType, params.DestinationType))
+}
+
+// IdentityConvertComplexOptionalProperty handles conversion for optional properties with complex elements
+// This function generates code that looks like this:
+// 	if <source> != nil {
+//		<code for producing result from destinationType.Element()>
+//		<destination> = &<result>
+//	}
+func IdentityConvertComplexOptionalProperty(builder *ConversionFunctionBuilder, params ConversionParameters) []dst.Stmt {
+	destinationType, ok := params.DestinationType.(*OptionalType)
+	if !ok {
+		return nil
+	}
+
+	sourceType, ok := params.SourceType.(*OptionalType)
+	if !ok {
+		return nil
+	}
+
+	// TODO: Dislike Typed suffix here -- change
+	tempVarIdent := builder.IdFactory.CreateIdentifier(params.NameHint+"Typed", NotExported)
+
+	newSource := &dst.UnaryExpr{
+		X:  params.ClonedSource(),
+		Op: token.MUL,
+	}
+
+	innerStatements := builder.BuildConversion(
+		params.WithDestination(dst.NewIdent(tempVarIdent)).
```

Comment:
I think this would be clearer if you just construct a new `ConversionParameters` and explicitly copy some fields from `params` and set some with new values. Otherwise you have to refer back to the struct def to work out which fields are the same and which are different.

Reply:
Done